### PR TITLE
Update Project Cache docs to reflect new functionality

### PR DESCRIPTION
**Pull request by dfederm**
Update Project Cache docs to reflect new functionality

Most the new stuff was added in #9214, but some other changes have gone in before then as well which made these docs out of date.

### CODE REVIEW

File: documentation/specs/project-cache.md [hunk 1]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+MSBuild will then build the project normally but under a detoured worker node. Because of this, the plugin will recieve `HandleFileAccess` and `HandleProcess` events. In this example implementation we will ignore `HandleProcess`. For `HandleFileAccess`, the plugin will simply store all `FileAccessData`s for a `FileAccessContext` to build up a list of all file accesses during the build. The plugin may decide to avoid storing the entire `FileAccessData` and instead just peel off the data it finds relevant (eg. paths, whether it was a read or write, etc).
+
+Once MSBuild is done building the project, it will call the plugin's `HandleProjectFinishedAsync`. Now the plugin knows the project is done and can process the results and add them a cache. In general it's only useful to cache successful results, so the plugin should filter out non-success results. The `FileAccessContext` provided can then be used to retrieve the list of `FileAccessData` the plugin recieved. These `FileAccessData` can be processed to understand which files were read and writted as part of the build.
```
Comment (DmitriyShepelev) [inline documentation/specs/project-cache.md:174]: Nit: add them **to** a cache

Comment (DmitriyShepelev) [reply documentation/specs/project-cache.md:174]: Nit; **written**

File: documentation/specs/project-cache.md [hunk 2]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+This can then be used for future builds.
+
+ ## Second Build (cache hits)
+ 
+ In the second build we have a populated cache and so it could be possible to get cache hits.
+
+ For a given project, `GetCacheResultAsync` will be invoked. The plugin can fingerprint the request and use that fingerprint to look up in its cache. If the cache entry exists, it can declare a cache hit.
+
+In the example above, if all inputs are the same as in the first build, we should end up with a fingerprint `F`. We look up in the metadata part of the cache (file `metadata/F`) and find that it exists. This means we have a cache hit. We can fetch that metadata `M` from the cache and find that is describes the output with path `O` and hash `H`. The plugin would then copy `content/H` to `O` and return the deserialized `BuildResult R` contained in `M` to MSBuild.
```
Comment (DmitriyShepelev) [inline documentation/specs/project-cache.md:192]: Nit: **it** describes

File: documentation/specs/project-cache.md [hunk 3]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+      - Potential Issue: serialization format may change between writing and reading the `BuildResult`, especially if binary serialization is used.
+
+## Cache add scenario
+
+- Upon a cache miss, MSBuild will generally handle a request as normal, ie by building it.
+- To facilitate the plugin being able to hande future builds, MSBuild can report file accesses as well as the build result for the BuildRequestData.
+- MSBuild disables the in-proc node and uses [Detours](https://github.com/microsoft/Detours) to observe file accesses of the worker nodes. It forwards this information to the plugin for it to use as desired.
```
Comment (DmitriyShepelev) [inline documentation/specs/project-cache.md:68]: Should there be a statement/section describing the performance cost of Detours/cache add overall?

**Review by DmitriyShepelev — APPROVED** [review]

File: documentation/specs/project-cache.md [hunk 4]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+  - This functionality has some implementation restrictions so will require additional opt-in. Specifically, the `/ReportFileAccesses` command-line flag or by setting `BuildParameters.ReportFileAccesses` for programatic use of `BuildManager`. If this is not set, no file accesses will be reported to the plugin, however the plugin will still be notified of the build result.
```
Comment (JanKrivanek) [inline documentation/specs/project-cache.md:69]: Those 3 points seem a bit redundant, while each telling  something slightly different ('MSBuild can report file accesses', 'MSBuild forwards file accesses info to the plugin', 'Unless opted in, no file accesses are reported') - wouldn't it be better to mention it just once, while mentioning the opt-in mechanics?

File: documentation/specs/project-cache.md [hunk 5]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+public abstract Task EndBuildAsync(PluginLoggerBase logger, CancellationToken cancellationToken);
+
+public abstract Task<CacheResult> GetCacheResultAsync(BuildRequestData buildRequest, PluginLoggerBase logger, CancellationToken cancellationToken);
+
+public virtual void HandleFileAccess(FileAccessContext fileAccessContext, FileAccessData fileAccessData);
+
+public virtual void HandleProcess(FileAccessContext fileAccessContext, ProcessData processData);
+
+public virtual Task HandleProjectFinishedAsync(FileAccessContext fileAccessContext, BuildResult buildResult, PluginLoggerBase logger, CancellationToken cancellationToken);
+```
```
Comment (JanKrivanek) [inline documentation/specs/project-cache.md:90]: This feels superfluous to the link to iplementation provided

Comment (JanKrivanek) [reply documentation/specs/project-cache.md:90]: Also it should probably link the `Deteours` section below (and leave the implementation details info - like the need to disable in-proc node etc. - to be mentioned only in that section. Now some info is duplicated)

File: documentation/specs/project-cache.md [hunk 6]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+Note that the Scheduler node cannot use Detours on itself, so the in-proc node is disabled when repoting file accesses. Additionally task yielding is disabled since it would leave to improperly associated file accesses.
+
+## Pipe synchronization
+
+Because the Detours implementation being used communicates over a pipe, and node communication is also over a pipe, and pipes are async, there is some coordination required for ensuring that file accesses are associated with the proper build request. For example, if a "project finished" signal comes through the node communication pipe, but the detours pipe still has a queue of file accesses which have not been processed yet, those file accesses might be processed after the worker node has moved onto some other project.
```
Comment (JanKrivanek) [inline documentation/specs/project-cache.md:134]:
```suggestion
Because the Detours implementation being used communicates over a pipe, and nodes communicate over a pipe as well, and pipes are async, there is some coordination required to ensure that file accesses are associated with the proper build request. For example, if a "project finished" signal comes through the node communication pipe, but the detours pipe still has a queue of file accesses which have not been processed yet, those file accesses might be processed after the worker node has moved onto some other project.
```

But I'm not a native speaker though

**Review by JanKrivanek — APPROVED** [review]
Overall it's great to have this documented in detail and with samples!

The document delivers all the necessary info - in some minor cases it could have been more clear and/or less redundant. But those are rather minor things - even as is now, it is great value added.

File: documentation/specs/project-cache.md [hunk 7]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+- The workaround logic activates only when MSBuild detects that it's running under VS.
+- When VS evaluates projects via `new Project(..)` (it does this on all the solution projects on solution load), the evaluator will search for and store all detected plugins in a static field on the `BuildManager`.
+- The first build request will check the static state for the presence of plugins. If there's a plugin, it will initialize it at that point.
+- Plugins will be given the graph entry points instead of the entire graph in this scenario.
+- There is currently no way to enable cache add scenarios in Visual Studio.
+
+# Detours (cache add scenario)
+
+In order for MSBuild to observe the file accesses as part of the build, it uses Detours on the worker nodes. In this way the Scheduler node will events for all file accesses done by the worker nodes. As the Scheduler knows what build request a worker node is working on at any given moment, it is able to properly associate the file access with a build request and dispatch these augmented events to plugins via the plugins' `HandleFileAccess` and `HandleProcess` implementations.
```
[the remark this answers is not in the document]
Comment (dfederm) [reply documentation/specs/project-cache.md:128]: Should be "will emit events"

File: documentation/specs/project-cache.md [hunk 8]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+## Pipe synchronization
+
+Because the Detours implementation being used communicates over a pipe, and node communication is also over a pipe, and pipes are async, there is some coordination required for ensuring that file accesses are associated with the proper build request. For example, if a "project finished" signal comes through the node communication pipe, but the detours pipe still has a queue of file accesses which have not been processed yet, those file accesses might be processed after the worker node has moved onto some other project.
+
+To address this problem, when a worker node finishes a project it will emit a dummy file access with a specific format known to MSBuild. When the scheduler not recieves as "project finished" event over the node communication pipe, it will wait to determine that the project is actually finished until it also recieves the dummy file access. This ensures that the all file accesses associated with the project have fully flushed from the pipe before the scheduler deterines the project is finished and schedules new work to the worker node (which would trigger new file accesses).
```
[the remark this answers is not in the document]
Comment (dfederm) [reply documentation/specs/project-cache.md:136]: "scheduler node receives"

I also spelled receives wrong. Twice.

File: documentation/specs/project-cache.md [hunk 9]
```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+- Only the user provided top level build requests are checked against the cache. The build requests issued recursively from the top level requests are not checked against the cache, since it is assumed that users issue build requests in reverse toposort order. Therefore when a project builds its references, those references should have already been built and present in MSBuild's internal cache, provided either by the project cache plugin or real builds. In other words, projects which are not well-described in the graph (eg using `<MSBuild>` tasks directly) will not benefit from the cache.
+- The plugin instance will get called in reverse topo sort order (from dependencies up towards dependents). This happens based on the requirement of using `/graph`. Building in reverse topo sort order is common between Visual Studio solution builds and higher build engines.
+- Plugins can function with and without a static graph. When a static graph is not provided, hints about the graph entry points are provided (details in Defining the "graph" when static graph is not available).
```
[the remark this answers is not in the document]
Comment (dfederm) [reply documentation/specs/project-cache.md:44]: Reworked some of these points.

File: documentation/specs/project-cache.md (hunk 4 above)
Comment (dfederm) [reply documentation/specs/project-cache.md:69]: Reworked some of the verbiage

File: documentation/specs/project-cache.md (hunk 5 above)
Comment (dfederm) [reply documentation/specs/project-cache.md:90]: Removing the signatures and linked to the `Plugin implementation guidance and simple example design` section.

File: documentation/specs/project-cache.md (hunk 6 above)
Comment (dfederm) [reply documentation/specs/project-cache.md:134]: That's much less awkward sentence structure :)

File: documentation/specs/project-cache.md [hunk 10]
```diff
@@ -14,106 +14,185 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+It can make sense for a fingerprint to be a hash of its inputs, so effectively is a byte array which can be represented by a string.
+
+At the beginning of the build, the plugin's `BeginBuildAsync` method will be called. As part of the `CacheContext`, the plugin is either given the graph or the entry points to the graph for which it can create a graph from. The plugin can use this graph to do some initial processing, like predicting various inputs which a project is likely to use. This information can then be stored to help construct a fingerprint for a build request later.
+
+## Cache storage
+
+Any storage mechanism can be used as a cache implementation, for example [Azure Blob Storage](https://azure.microsoft.com/en-us/products/storage/blobs/), or even just the local filesystem. At least in this example the only real requirement is that it can be used effectively as a key-value store. In many cases it can be useful for content to be keyed by its hash, and for the metadata file to be keyed by the fingerprint. In particular when content is keyed by hash, it is effectively deduplicated across multiple copies of the same file, which is common in builds.
```
Comment (rainersigwald) [inline documentation/specs/project-cache.md:152]:
```suggestion
Any storage mechanism can be used as a cache implementation, for example [Azure Blob Storage](https://azure.microsoft.com/products/storage/blobs/), or even just the local filesystem. At least in this example the only real requirement is that it can be used effectively as a key-value store. In many cases it can be useful for content to be keyed by its hash, and for the metadata file to be keyed by the fingerprint. In particular when content is keyed by hash, it is effectively deduplicated across multiple copies of the same file, which is common in builds.
```

File: documentation/specs/project-cache.md [hunk 11]
```diff
@@ -14,106 +14,185 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+For illustration purposes, consider our cache implementation is based on a simple filesystem which a separate metadata and content directory inside it. Under the metadata dir, each file is a metadata file where the filename matches the fingerprint it's describing. Under the content dir, each file is a content file where the filename matches the hash of the content itself.
```
Comment (rainersigwald) [inline documentation/specs/project-cache.md:154]:
```suggestion
For illustration purposes, consider our cache implementation is based on a simple filesystem with a separate metadata and content directory inside it. Under the metadata dir, each file is a metadata file where the filename matches the fingerprint it's describing. Under the content dir, each file is a content file where the filename matches the hash of the content itself.
```

File: documentation/specs/project-cache.md [hunk 12]
```diff
@@ -14,106 +14,185 @@ This change also simplifies and unifies user experiences. MSBuild works the same
...
+
+In our example, we can use the read files to construct a fingerprint for the build request. We can then add the files written during the build ("outputs") to some cache implementation.
+
+The plugin would then create some metadata describing the outputs (eg. the paths and hashes) and the serialized `BuildResult`, and associate it with the fingerprint and put that assocation in the cache.
+
+To illustrate this, consider a project with fingerprint `F` which wrote a single file `O` with hash `H` and had `BuildResult R`. The plugin could create a metadata file `M` which describes the outputs of the build (the path and hash of `O`) as well as the serialized `R`. Using the cache implementation described above, the plugin would write the following two files to the cache:
+ - `metadata/F -> M`
+ - `content/H -> O`
```
Comment (rainersigwald) [inline documentation/specs/project-cache.md:172]: I found this kinda hard to follow, I was expecting R to appear in one of the files. It's not _wrong_ now that I read it a few times but if you have an easy rephrase I'd appreciate it. Maybe something like
```suggestion
To illustrate this, consider a project with fingerprint `F` which wrote a single file `O` with hash `H` and had `BuildResult R`. The plugin could create a metadata file `M` which describes the outputs of the build (the path and hash of `O`) as well as the serialized `R`. Using the cache implementation described above, the plugin would write the following two files to the cache:
 - `metadata/F -> M:'{outputs: {path: 'path/to/O', hash: H}, result: R}`
 - `content/H -> O`
```

**Review by rainersigwald — APPROVED** [review]
A few minor tweak suggestions but looks great to me, thanks and sorry for the review delay!
